File: Project 4/Project_4a.f90
!Final project part 1
!Module for flow simulations of liquid through tube
!This module contains a few module variables (see comments below)
!and four subroutines:
!jacobi: Uses jacobi iteration to compute solution
! to flow through tube
!sgisolve: To be completed. Use sgi method to
! compute flow through tube
!mvec: To be completed; matrix-vector multiplication z = Ay
!mtvec: To be completed; matrix-vector multiplication z = A^T y
module flow
    implicit none
    real(kind=8), parameter :: pi = acos(-1.d0)
    integer :: numthreads !number of threads used in parallel regions
    integer :: fl_kmax=100000 !max number of iterations
    real(kind=8) :: fl_tol=0.00000001d0 !convergence criterion
    real(kind=8), allocatable, dimension(:) :: fl_deltaw !|max change in w| each iteration
    real(kind=8) :: fl_s0=0.1d0 !deformation magnitude

contains
!-----------------------------------------------------
!Solve 2-d tube flow problem with Jacobi iteration
subroutine jacobi(n,w)
    !input  n: number of grid points (n+2 x n+2) grid
    !output w: final velocity field
    !Should also compute fl_deltaw(k): max(|w^k - w^k-1|)
    !A number of module variables can be set in advance.

    integer, intent(in) :: n
    real(kind=8), dimension(0:n+1,0:n+1), intent(out) :: w
    integer :: i1,j1,k1
    real(kind=8) :: del_r,del_t,del_r2,del_t2
    real(kind=8), dimension(0:n+1) :: s_bc,fac_bc
    real(kind=8), dimension(0:n+1,0:n+1) :: r,r2,t,RHS,w0,wnew,fac,fac2,facp,facm

    if (allocated(fl_deltaw)) then
      deallocate(fl_deltaw)
    end if
    allocate(fl_deltaw(fl_kmax))


    !grid--------------
    del_t = 0.5d0*pi/dble(n+1)
    del_r = 1.d0/dble(n+1)
    del_r2 = del_r**2
    del_t2 = del_t**2


    do i1=0,n+1
        r(i1,:) = i1*del_r
    end do

    do j1=0,n+1
        t(:,j1) = j1*del_t
    end do
    !-------------------

    !Update-equation factors------
    r2 = r**2
    fac = 0.5d0/(r2*del_t2 + del_r2)
    facp = r2*del_t2*fac*(1.d0+0.5d0*del_r/r) !alpha_p/gamma
    facm = r2*del_t2*fac*(1.d0-0.5d0*del_r/r) !alpha_m/gamma
    fac2 = del_r2 * fac !beta/gamma
    RHS = fac*(r2*del_r2*del_t2) !1/gamma
    !----------------------------

    !set initial condition/boundary deformation
    w0 = (1.d0-r2)/4.d0
    w = w0
    wnew = w0
    s_bc = fl_s0*exp(-10.d0*((t(0,:)-pi/2.d0)**2))/del_r
    fac_bc = s_bc/(1.d0+s_bc)


    !Jacobi iteration
    do k1=1,fl_kmax
        wnew(1:n,1:n) = RHS(1:n,1:n) + w(2:n+1,1:n)*facp(1:n,1:n) + w(0:n-1,1:n)*facm(1:n,1:n) + &
                                         (w(1:n,0:n-1) + w(1:n,2:n+1))*fac2(1:n,1:n)

        !Apply boundary conditions
        wnew(:,0) = wnew(:,1) !theta=0
        wnew(:,n+1) = wnew(:,n) !theta=pi/2
        wnew(0,:) = wnew(1,:) !r=0
        wnew(n+1,:) = wnew(n,:)*fac_bc !r=1s

        fl_deltaw(k1) = maxval(abs(wnew-w)) !compute relative error

        w=wnew    !update variable
        if (fl_deltaw(k1)<fl_tol) exit !check convergence criterion
        if (mod(k1,1000)==0) print *, k1,fl_deltaw(k1)
    end do


  !  print *, 'k,error=',k1,fl_deltaw(min(k1,fl_kmax))

end subroutine jacobi
!-----------------------------------------------------

!Solve 2-d tube flow problem with sgi method
subroutine sgisolve(n,w)
  !input  n: number of grid points (n+2 x n+2) grid
  !output w: final velocity field stored in a column vector
  !Should also compute fl_deltaw(k): max(|w^k - w^k-1|)
  !A number of module variables can be set in advance.
  use omp_lib
  integer, intent(in) :: n
  real(kind=8), dimension((n+2)*(n+2)), intent(out) :: w
  real(kind=8) :: del_t,del_r
  !add other variables as needed
  integer :: i1,j1
  real(kind=8), dimension((n+2)*(n+2)) :: x_old,x_new,d,e,b
  real(kind=8) :: n1,step_size,meu,w_tol
  real(kind=8) :: del_r2,del_t2,e_newsum,e_oldsum,Ad_sum
  real(kind=8), dimension(n+2) :: s_bc,fac_bc,r,t,fac,fac2,facp,facm,r2,RHS
  real(kind=8), dimension((n+2)*(n+2)) ::Ad,Md

  if (allocated(fl_deltaw)) then
    deallocate(fl_deltaw)
  end if
  allocate(fl_deltaw(fl_kmax))

  !number of threads
    !$ call omp_set_num_threads(numthreads)

    del_t = 0.5d0*pi/dble(n+1)
    del_r = 1.d0/dble(n+1)
    del_r2 = del_r**2
    del_t2 = del_t**2


    do i1=0,n+1
        r(i1+1) = i1*del_r
        t(i1+1) = i1*del_t
    end do

    !-------------------

    !Update-equation factors------
    r2 = r**2
    fac = 0.5d0/(r2*del_t2 + del_r2)
    facp = r2*del_t2*fac*(1.d0+0.5d0*del_r/r) !alpha_p/gamma
    facm = r2*del_t2*fac*(1.d0-0.5d0*del_r/r) !alpha_m/gamma
    fac2 = del_r2 * fac !beta/gamma
    RHS = fac*(r2*del_r2*del_t2) !1/gamma

    s_bc = fl_s0*exp(-10.d0*((t(:)-pi/2.d0)**2))/del_r
    fac_bc = s_bc/(1.d0+s_bc)

  !set dummy variables
  n1=(n+2)*(n+2)
  !work out b
  b=0.d0
  !$OMP parallel do private(j1)
  do i1=1,n
    do j1=2,n+1
      b(i1*(n+2)+j1)=-RHS(i1+1)
    end do
  end do
  !$OMP end parallel do

  call mtvec(n,fac,fac2,facp,facm,fac_bc,b,d)

!set intial steps in sig solve
  x_old=0.d0
  e=d
  e_oldsum=0.d0

  !$OMP parallel do reduction(+:e_oldsum)
  do j1=1,(n+2)*(n+2)
    e_oldsum=(e(j1))**2+e_oldsum
  end do
  !$OMP end parallel do
  w_tol=fl_tol+1
  i1=0
  !iterate until max tol is reached
  do while  (w_tol>fl_tol)
    !work outstep size
    call mvec(n,fac,fac2,facp,facm,fac_bc,d,Ad)

    !workout dot product of Ad
    Ad_sum=0.d0
    !$OMP parallel do reduction(+:Ad_sum)
    do j1=1,(n+2)*(n+2)
      Ad_sum=(Ad(j1))**2+Ad_sum
    end do
    !$OMP end parallel do

    !workout step size
    step_size=e_oldsum/Ad_sum
    !work estimated solution
    x_new=x_old+step_size*d

    !work out resdual by using mvec result above and mtvec
    call mtvec(n,fac,fac2,facp,facm,fac_bc,Ad,Md)

    !work out new residual
    e=e-step_size*Md

    !work magnitude of new search direction
    e_newsum=0.d0
    !$OMP parallel do reduction(+:e_newsum)
    do j1=1,(n+2)*(n+2)
      e_newsum=(e(j1))**2+e_newsum
    end do
    !$OMP end parallel do

    meu=e_newsum/e_oldsum
    !work out new d
    d=e+meu*d

    !work out max change in iteration
    w_tol=maxval(abs(x_old-x_new))
    !update variables
    x_old=x_new
    e_oldsum=e_newsum
    !record what time step we are on
    i1=i1+1
    !store tolerence
    fl_deltaw(i1)=w_tol

    !have a max number of iterations
    if (i1>fl_kmax) then
      w_tol=0
      print *, "failed to converge"
    end if

  end do

!output result
  w=x_new


end subroutine sgisolve


!Compute matrix-vector multiplication, z = Ay
subroutine mvec(n,fac,fac2,facp,facm,fac_bc,y,z)
    !input n: grid is (n+2) x (n+2)
    ! fac,fac2,facp,facm,fac_bc: arrays that appear in
    !   discretized equations
    ! y: vector multipled by A
    !output z: result of multiplication Ay
    use omp_lib
    implicit none
    integer, intent(in) :: n
    integer :: i1,j1,n1,n2
    real(kind=8), dimension(n+2), intent(in) :: fac,fac2,facp,facm,fac_bc
    real(kind=8), dimension((n+2)*(n+2)), intent(in) :: y
    real(kind=8), dimension((n+2)*(n+2)), intent(out) :: z
    !$ call omp_set_num_threads(numthreads)

    !add other variables as needed
    n1=(n+2)*(n+2)

    do i1=1,n+2
      !boundary condtion for r =1
      !should be other wat round
      z(i1)=y(i1+n+2)-y(i1)
      !boundary at centre
      z((n+2)*(n+1)+i1)=y((n+2)*(n+1)+i1)-fac_bc(i1)*y((n+2)*(n)+i1)
    end do

    !$OMP parallel do private(n2,j1)
    do i1=1,n
      !dummy variable to improve effieceeny
      n2=i1*(n+2)+1
      !boundary condition theta 1
      z(n2)=y(n2+1)-y(n2)
      !boundary condition theta 2
      z(n2+n+1)=y(n2+n+1)-y(n2+n)
      do j1=1,n
        z(j1+n2)=-y(j1+n2)+fac2(i1+1)*(y(j1-1+n2)+y(j1+1+n2))+facm(i1+1)*y(j1+n2-n-2)+facp(i1+1)*y(j1+n2+n+2)
      end do
    end do
    !$OMP end parallel do
end subroutine mvec


!Compute matrix-vector multiplication, z = A^T y
subroutine mtvec(n,fac,fac2,facp,facm,fac_bc,y,z)
    !input n: grid is (n+2) x (n+2)
    ! fac,fac2,facp,facm,fac_bc: arrays that appear in
    !   discretized equations
    ! y: vector multipled by A^T
    !output z: result of multiplication A^T y
    use omp_lib
    implicit none
    integer, intent(in) :: n
    integer :: i1,j1,n1,n2,n3
    real(kind=8), dimension(n+2), intent(in) :: fac,fac2,facp,facm,fac_bc
    real(kind=8), dimension((n+2)*(n+2)), intent(in) :: y
    real(kind=8), dimension((n+2)*(n+2)), intent(out) :: z
    !$ call omp_set_num_threads(numthreads)


    !add other variables as needed
    n1=(n+2)*(n+2)
    n3=(n+2)*(n+1)
    !set first and last element of first sub matrix
    z(1)=-y(1)
    z(n+2)=-y(n+2)

    !work out first n+2 rows
    do i1=2,n+1
      !boundary condtion for r =1
      z(i1)=facm(2)*y(i1+n+2)-y(i1)
    end do


    !dummy variable to improve effieceeny
    n2=(n+2)+1
    !boundary condition theta 1
    z(n2)=fac2(2)*y(n2+1)-y(n2)+y(n2-n-2)
    !first row of sub diagonal matrix effect by boundary condition 1
    z(n2+1)=y(n2)-y(n2+1)+fac2(2)*y(n2+2)+facm(3)*y(n2+n+3)+y(n2-n-1)
    !boundary condition theta 2 last row of sub matrix
    z(n2+n+1)=y(n2+n+1)+fac2(2)*y(n2+n)+y(n2-1)

    do j1=2,n-1
      z(j1+n2)=-y(j1+n2)+fac2(2)*(y(j1-1+n2)+y(j1+1+n2))+facm(3)*y(j1+n2+n+2)+y(j1+n2-n-2)
    end do

    !2nd to last row of sub diagonal matrix effect by boundary condition 2
    z(n2+n)=fac2(2)*y(n2+n-1)-y(n2+n)-y(n2+n+1)+facm(3)*y(n2+n+n+2)+y(n2-2)


    !$OMP parallel do private(j1,n2)
    do i1=2,n-1
      !dummy variable to improve effieceeny
      n2=i1*(n+2)+1
      !boundary condition theta 1
      z(n2)=fac2(i1+1)*y(n2+1)-y(n2)
      !first row of sub diagonal matrix effect by boundary condition 1
      z(n2+1)=y(n2)-y(n2+1)+fac2(i1+1)*y(n2+2)+facm(i1+2)*y(n2+n+3)+facp(i1)*y(n2-n-1)
      !boundary condition theta 2 last row of sub matrix
      z(n2+n+1)=y(n2+n+1)+fac2(i1+1)*y(n2+n)

      do j1=2,n-1
        z(j1+n2)=-y(j1+n2)+fac2(i1+1)*(y(j1-1+n2)+y(j1+1+n2))+facm(i1+2)*y(j1+n2+n+2)+facp(i1)*y(j1+n2-n-2)
      end do
      !2nd to last row of sub diagonal matrix effect by boundary condition 2
      z(n2+n)=fac2(i1+1)*y(n2+n-1)-y(n2+n)-y(n2+n+1)+facm(i1+2)*y(n2+n+n+2)+facp(i1)*y(n2-2)
    end do
    !$OMP end parallel do

!!!!Calculting 2nd to last submatric
    !dummy variable to improve effieceeny
    n2=n*(n+2)+1
    !boundary condition theta 1
    z(n2)=fac2(n+1)*y(n2+1)-y(n2)-fac_bc(1)*y(n2+n+2)
    !first row of sub diagonal matrix effect by boundary condition 1
    z(n2+1)=y(n2)-y(n2+1)+fac2(n+1)*y(n2+2)-fac_bc(2)*y(n2+n+3)+facp(n)*y(n2-n-1)
    !boundary condition theta 2 last row of sub matrix
    z(n2+n+1)=y(n2+n+1)+fac2(n+1)*y(n2+n)-fac_bc(n+2)*y(n2+n+n+3)

    do j1=2,n-1
      z(j1+n2)=-y(j1+n2)+fac2(n+1)*(y(j1-1+n2)+y(j1+1+n2))-fac_bc(j1+1)*y(j1+n2+n+2)+facp(n)*y(j1+n2-n-2)
    end do

    !2nd to last row of sub diagonal matrix effect by boundary condition 2
    z(n2+n)=fac2(n+1)*y(n2+n-1)-y(n2+n)-y(n2+n+1)-fac_bc(n+1)*y(n2+n+n+2)+facp(n)*y(n2-2)

      !last sub matrix
    do i1=2,n+1
      !boundary at centre
      z((n+2)*(n+1)+i1)=y((n+2)*(n+1)+i1)+facp(n+1)*y((n+2)*(n+1)+i1-n-2)
    end do
      !first and last entry of last sub matrix
      z((n+2)*(n+1)+1)=y((n+2)*(n+1)+1)
      z((n+2)*(n+2))=y((n+2)*(n+2))
    end subroutine mtvec



end module flow

File: Project 4/Project_4a.py
"""Final project, part 1"""
import numpy as np
import matplotlib.pyplot as plt
from m1 import flow as fl
import time

#assumes files have been compiled with belo command lind to get omp form and just seriel form
#f2py3 --f90flags='-fopenmp' -c p41.f90 -m m1 -lgomp
#f2py3 -c p41.f90 -m q1

def jacobi(n,kmax=10000,tol=1.0e-8,s0=0.1,display=False):
    """ Solve liquid flow model equations with
        jacobi iteration.
        Input:
            n: number of grid points in r and theta
            kmax: max number of iterations
            tol: convergence test parameter
            s0: amplitude of cylinder deformation
            display: if True, plots showing the velocity field and boundary deformation
            are generated
        Output:
            w,deltaw: Final velocity field and |max change in w| each iteration
    """

    #-------------------------------------------
    #Set Numerical parameters and generate grid
    Del_t = 0.5*np.pi/(n+1)
    Del_r = 1.0/(n+1)
    Del_r2 = Del_r**2
    Del_t2 = Del_t**2
    r = np.linspace(0,1,n+2)
    t = np.linspace(0,np.pi/2,n+2) #theta
    tg,rg = np.meshgrid(t,r) # r-theta grid

    #Factors used in update equation (after dividing by gamma)
    rg2 = rg*rg
    fac = 0.5/(rg2*Del_t2 + Del_r2)
    facp = rg2*Del_t2*fac*(1+0.5*Del_r/rg) #alpha_p/gamma
    facm = rg2*Del_t2*fac*(1-0.5*Del_r/rg) #alpha_m/gamma
    fac2 = Del_r2*fac #beta/gamma
    RHS = fac*(rg2*Del_r2*Del_t2) #1/gamma

    #set initial condition/boundary deformation
    w0 = (1-rg**2)/4 #Exact solution when s0=0
    s_bc = s0*np.exp(-10.*((t-np.pi/2)**2))/Del_r
    fac_bc = s_bc/(1+s_bc)

    deltaw = []
    w = w0.copy()
    wnew = w0.copy()

    #Jacobi iteration
    for k in range(kmax):
        #Compute wnew
        wnew[1:-1,1:-1] = RHS[1:-1,1:-1] + w[2:,1:-1]*facp[1:-1,1:-1] + w[:-2,1:-1]*facm[1:-1,1:-1] + (w[1:-1,:-2] + w[1:-1,2:])*fac2[1:-1,1:-1] #Jacobi update

        #Apply boundary conditions
        wnew[:,0] = wnew[:,1] #theta=0
        wnew[:,-1] = wnew[:,-2] #theta=pi/2
        wnew[0,:] = wnew[1,:] #r=0
        wnew[-1,:] = wnew[-2,:]*fac_bc #r=1s

        #Compute delta_p
        deltaw += [np.max(np.abs(w-wnew))]
        w = wnew.copy()
        if k%1000==0: print("k,dwmax:",k,deltaw[k])
        #check for convergence
        if deltaw[k]<tol:
            print("Converged,k=%d,dw_max=%28.16f " %(k,deltaw[k]))
            break

    deltaw = deltaw[:k+1]

    if display:
        #plot final velocity field, difference from initial guess, and cylinder
        #surface
        plt.figure()
        plt.contour(t,r,w,50)
        plt.xlabel(r'$\theta$')
        plt.ylabel('r')
        plt.title('Final velocity field')

        plt.figure()
        plt.contour(t,r,np.abs(w-w0),50)
        plt.xlabel(r'$\theta$')
        plt.ylabel('r')
        plt.title(r'$|w - w_0|$')

        plt.figure()
        plt.polar(t,np.ones_like(t),'k--')
        plt.polar(t,np.ones_like(t)+s_bc*Del_r,'r-')
        plt.title('Deformed cylinder surface')

    return w,deltaw



def performance():
    """Analyze performance of codes
    Add input/output variables as needed.
    """

    return None



if __name__=='__main__':
    #Add code below to call performance
    #and generate figures you are submitting in
    #your repo.
    marray=[10,20,30,40,50,60,70,80,90,100,150,200,300,400,500]
    m_len=len(marray)
    #set up data for time entries m varying
    time_ser=np.zeros(m_len)
    time_par2=np.zeros(m_len)
    time_par4=np.zeros(m_len)
    from q1 import flow as fl

    for i in range(m_len):
            start_ser_fortran = time.time()
            w1 = fl.sgisolve(marray[i])
            end_ser_fortran = time.time()
            time_ser[i]=end_ser_fortran - start_ser_fortran
    from m1 import flow as fl
    fl.numthreads=2
    for i in range(m_len):
            start_ser_fortran = time.time()
            w1 = fl.sgisolve(marray[i])
            end_ser_fortran = time.time()
            time_par2[i]=end_ser_fortran - start_ser_fortran

    fl.numthreads=4
    for i in range(m_len):
            start_ser_fortran = time.time()
            w1 = fl.sgisolve(marray[i])
            end_ser_fortran = time.time()
            time_par4[i]=end_ser_fortran - start_ser_fortran


            plt.figure()
            plt.plot(marray,time_ser/time_par2,label='2 threads')
            plt.plot(marray,time_ser/time_par4,label='4 threads')
            plt.plot(marray,time_ser*0+1)
            plt.legend()
            plt.legend()
            plt.xlabel("M value")
            plt.ylabel("run time (seconds)")
            plt.title("Matthew Cowley, Speed up for MPI with varying M Relative to Serial")


    marray=[10,20,30,40,50,60,70,80,90,100,110,120]
    m_len=len(marray)
    #set up data for time entries m varying
    time_ser=np.zeros(m_len)
    time_par1=np.zeros(m_len)
    time_python=np.zeros(m_len)
    time_par2=np.zeros(m_len)
    time_jack=np.zeros(m_len)

    fl.numthreads=1
    for i in range(m_len):
            start_ser_fortran = time.time()
            w1 = fl.sgisolve(marray[i])
            end_ser_fortran = time.time()
            time_par1[i]=end_ser_fortran - start_ser_fortran

    fl.numthreads=2
    for i in range(m_len):
            start_ser_fortran = time.time()
            w1 = fl.sgisolve(marray[i])
            end_ser_fortran = time.time()
            time_par2[i]=end_ser_fortran - start_ser_fortran

    for i in range(m_len):
            start_ser_fortran = time.time()
            w1 = fl.jacobi(marray[i])
            end_ser_fortran = time.time()
            time_jack[i]=end_ser_fortran - start_ser_fortran


    from q1 import flow as fl
    for i in range(m_len):
            start_ser_fortran = time.time()
            w1 = fl.sgisolve(marray[i])
            end_ser_fortran = time.time()
            time_ser[i]=end_ser_fortran - start_ser_fortran

    marray=[10,20,30,40,50,60,70,80,90,100,150,200,300,400,500]
    m_len=len(marray)
    #set up data for time entries m varying
    time_ser=np.zeros(m_len)
    time_par2=np.zeros(m_len)
    time_par4=np.zeros(m_len)
    from q1 import flow as fl
    for i in range(m_len):
            start_ser_fortran = time.time()
            w1 = fl.sgisolve(marray[i])
            end_ser_fortran = time.time()
            time_ser[i]=end_ser_fortran - start_ser_fortran
    from m1 import flow as fl
    fl.numthreads=2
    for i in range(m_len):
            start_ser_fortran = time.time()
            w1 = fl.sgisolve(marray[i])
            end_ser_fortran = time.time()
            time_par2[i]=end_ser_fortran - start_ser_fortran

    fl.numthreads=4
    for i in range(m_len):
            start_ser_fortran = time.time()
            w1 = fl.sgisolve(marray[i])
            end_ser_fortran = time.time()
            time_par4[i]=end_ser_fortran - start_ser_fortran


    plt.figure()
    plt.plot(marray,time_ser/time_par2,label='2 threads')
    plt.plot(marray,time_ser/time_par4,label='4 threads')
    plt.plot(marray,time_ser*0+1)
    plt.legend()
    plt.legend()
    plt.xlabel("Speed up")
    plt.ylabel("run time (seconds)")
    plt.title("Matthew Cowley, Speed up for MPI with varying M Relative to Seriel code")





#last figure

    from m1 import flow as fl
fl.numthreads=2
w1 = fl.sgisolve(300)
g=fl.fl_deltaw
w2=fl.jacobi(300)
f=fl.fl_deltaw

plt.figure()
plt.plot(np.arange(1,101,1),f[1:101],label='m=300, Jacobi')
plt.plot(np.arange(1,101,1),g[1:101],label='m=300, SGI 2 cores')
plt.legend()
plt.xlabel("iterations")
plt.ylabel("max change in velocity ")
plt.title("Matthew Cowley, How Jacobi and SGI converge")
